doc/source/development/rfc/rfc101_raster_dataset_threadsafety.rst

```diff
@@ -0,0 +1,484 @@
+Target:        GDAL 3.10
+============== =============================================
+
+Summary
+-------
+
+This RFC enables users to get instances of :cpp:class:`GDALDataset`
+(and their related objects such as :cpp:class:`GDALRasterBand`) that are
+thread-safe for read-only raster operations, that is such instances can
+be safely used from multiple threads without locking.
+
+Terminology
+-----------
+
+The exact meaning of the terms ``thread-safe`` or ``re-entrant`` is not fully
+standardized. We will use here the `QT definitions <https://doc.qt.io/qt-5/threads-reentrancy.html>`__.
+In particular, a C function or C++ method is said to be re-entrant if it can
+be called simultaneously from multiple threads, *but* only if each invocation
+uses its own data/instance. On the contrary, it is thread-safe is if can be
+called on the same data/instance (so thread-safe is stronger than re-entrant)
+
+Motivation
+----------
+
+A number of raster algorithms can be designed to read chunks of a raster in
+an independent and concurrent way, with resulting speed-ups when using
+multi-threading. Currently, given a GDALDataset instance is not thread-safe,
+this requires either to deal with I/O in a single thread, or through a mutex
+to protect against concurrent use, or one needs to open a separate GDALDataset
+for each worker thread. Both approaches can complicate the writing of such
+algorithms. The enhancement of this RFC aims at providing a special GDALDataset
+instance that can be used safely from multiple threads. Internally, it does use
+one GDALDataset per thread, but hides this implementation detail to the user.
+
+C and C++ API extensions
+------------------------
+
+A new ``GDAL_OF_THREAD_SAFE`` opening flag is added to be specified to
+:cpp:func:`GDALOpenEx` / :cpp:func:`GDALDataset::Open`. This flag is for now
+mutually exclusive with ``GDAL_OF_VECTOR``, ``GDAL_OF_MULTIDIM_RASTER`` and
+``GDAL_OF_UPDATE``. That is this flag is only meant for read-only raster
+operations (``GDAL_OF_RASTER | GDAL_OF_THREAD_SAFE``).
+
+To know if a given dataset can be used in a thread-safe way, the following
+C++ method is added to the GDALDataset class:
+
+.. code-block:: c++
+
+    /** Return whether this dataset, and its related objects (typically raster
+     * bands), can be called for the intended scope.
+     *
+     * Note that in the current implementation, nScopeFlags should be set to
+     * GDAL_OF_RASTER, as thread-safety is limited to read-only operations and
+     * excludes operations on vector layers (OGRLayer) or multidimensional API
+     * (GDALGroup, GDALMDArray, etc.)
+     *
+     * This is the same as the C function GDALDatasetIsThreadSafe().
+     *
+     * @since 3.10
+     */
+    bool IsThreadSafe(int nScopeFlags) const;
+
+
+The corresponding C function is added:
+
+.. code-block:: c
+
+    bool GDALDatasetIsThreadSafe(GDALDatasetH hDS, int nScopeFlags,
+                                 CSLConstList papszOptions);
```

Comment:
The C++ version could easily be extended in a backwards compatible way by adding an optional parameter if we need it, but that would break the C API, and this provision to avoid having to do a GDALDatasetIsThreadSafeEx() in the future